test/e2e/all_in_one_test.go

```diff
@@ -317,8 +324,15 @@ func verifyIngress(ctx context.Context, t *testing.T, env environments.Environme
 
 func verifyEnterprise(ctx context.Context, t *testing.T, env environments.Environment, adminPassword string) {
 	t.Log("finding the ip address for the admin API")
-	service, err := env.Cluster().Client().CoreV1().Services(namespace).Get(ctx, "kong-admin", metav1.GetOptions{})
-	require.NoError(t, err)
+	service, err := env.Cluster().Client().CoreV1().Services(namespace).Get(ctx, adminServiceName, metav1.GetOptions{})
```

Comment:
We do not have "kong-admin" service exposed under https://github.com/Kong/kubernetes-ingress-controller/blob/main/deploy/single-v2/all-in-one-enterprise-dbless.yaml, how can this verification ?

Reply:
This is a non-issue: please see the lines underneath this one, there is logic to create it if it doesn't exist. This makes it generically usable whether you use dbless OR postgres (which does have it).

Reply:
The flow is  no "kong-admin" service in the manifest, and
deploy the manifest; 
check non existing "kong-admin" service; 
create the "kong-admin" service

to check a resource that was not deployed, that does not sound logically.

Reply:
> to check a resource that was not deployed, that does not sound logically.

I'm unfortunately really unclear by what you mean, and I'm also unclear on what action you would like to see taken.

This code currently works and is used in two places:

- `TestDeployAllInOneEnterprisePostgres`
- `TestDeployAllInOneEnterpriseDBLESS`

And each logic path is taken:

- `TestDeployAllInOneEnterprisePostgres` already has a `kong-admin` service, so no action is taken
- `TestDeployAllInOneEnterpriseDBLESS` has **no** `kong-admin` service, so it is created

The result is that during the tests either deployment's Kong Admin API is accessible for testing. This works functionality as seen by the test results.

Can you let me know if that helps to clear anything up? Also can you provide a suggestion if you still don't consider this resolved?

Reply:
What I mean following lines  https://github.com/Kong/kubernetes-ingress-controller/blob/3cdcf35e1ffe4332ca17f767cfaa0d81637b7962/test/e2e/all_in_one_test.go#L109-L139 does not deploy any kong-admin service based on the manifest (deploy/single-v2/all-in-one-enterprise-dbless.yaml) you are using 

So the verify enterprise https://github.com/Kong/kubernetes-ingress-controller/blob/3cdcf35e1ffe4332ca17f767cfaa0d81637b7962/test/e2e/all_in_one_test.go#L142
does not need check kong-admin service.

I know it's nit pick, but it is a logic flaw.

Reply:
> I know it's nit pick, but it is a logic flaw.

I'm not understanding your perspective. If you're suggesting that I remove the GET logic and only create the service, this will cause the postgres tests to fail because they already provide their own `kong-admin` service.

Reply:
Since the kong-admin service does not exist, and create it directly, why would that cause postgress tests failure ?  Could you please help me to understand the case here ?

Otherwise, this PR LGTM, ready to go ! 👍

Reply:
> why would that cause postgress tests failure

The code currently _proactively creates_ the `kong-admin` service if it is missing, meaning that when the DBLESS tests use this code it is created, but when the postgres tests use this code it's not created and the existing one is used instead. Therefore if I were to remove the `env.Cluster().Client().CoreV1().Services(namespace).Get(` and _only_ do the create, the DBLESS tests would continue to succeed, but the postgres tests would fail with a 409 Conflict.

Reply:
The test case should use the setup which deployed in enterprise-dbless.yaml and enterprise-postgress.yaml, should not make any assumption. 
If we do not expose kong-admin service in dbeless mode, that means our user is not allowed to access the service directly, and we do not need test them.
If we expose kong-admin service in postgress mode, that means users need them, and we should test accordingly.

Reply:
> The test case should use the setup which deployed in enterprise-dbless.yaml and enterprise-postgress.yaml, should not make any assumption.

What assumption? Sorry I'm very confused by this statement.

>  If we do not expose kong-admin service in dbeless mode, that means our user is not allowed to access the service directly, and we do not need test them.

Check the remainder of the DBLESS test out, there's a _specific reason_ to expose the admin API: to check that the admin API is in fact reporting that it is `enterprise` so I would strongly argue that we _do_ want to access the service directly to verify that data point.

>  If we expose kong-admin service in postgress mode, that means users need them, and we should test accordingly.

I don't know if this comment was intended for some action to be taken, as this is already being tested in the code by way of testing workspaces.

Reply:
Question (non-blocker): verifyEnterprise(), checking out enterprise version for what ? https://github.com/Kong/kubernetes-ingress-controller/pull/1816/files#diff-d7850eda46771ac1eb42d866bb557fd98bdea88b6bcfba6b26b10cc975a40856L136 ,

 Based on above verification we are looking for some enterprise feature test cases since 
https://github.com/Kong/kubernetes-ingress-controller/pull/1816/files#diff-d7850eda46771ac1eb42d866bb557fd98bdea88b6bcfba6b26b10cc975a40856L130-L133
is exactly the same as Non-Enterprise DBLess https://github.com/Kong/kubernetes-ingress-controller/pull/1816/files#diff-d7850eda46771ac1eb42d866bb557fd98bdea88b6bcfba6b26b10cc975a40856R103-L100

Reply:
I'm really confused by the above question, but I'm gathering that the problem here has to do with the nature of _implicit_ creation of the `LoadBalancer` type `Service` causing some confusion when reading the code so I've opted to make it _explicit_ instead in hopes that this will resolve the confusion here:

d41e854c3bd4b819abc0a98411b9c7a23a0e9509

Let me know if this resolves your concerns here.

Reply:
I think the confusing is why verifyEnterprise() w/o a valid test case ? This can be converted into an AI - add a valid test case after verifyEnterprise().
In the test module,  both non-enterprise DBLess and enterprise DBLess have exactly same test steps - deploy, verifyIngress

Reply:
> I think the confusing is why verifyEnterprise() w/o a valid test case ? This can be converted into an AI - add a valid test case after verifyEnterprise().

I'm very confused, there are multiple test cases using this function? You say "without a valid test case" but both the dbless and postgres enterprise test cases use this function?

Can you provide a diff of the change you would like to see please? That might help me understand better what the requested change is.

Reply:
I think "a valid test case" mean that a test case (steps) that differentiate the underlying setup is kong enterprise DBLess, instead of non-enterprise DBLess, or enterprise postgres, so that we can see the value of verifyEnterprise() in the test module.

Reply:
The comment is not resolved. but this is CI blocker, mark as resolved to get the fix in.